builtin/k8s/task.go

```diff
@@ -0,0 +1,271 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	k8sresource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+)
+
+// TaskLauncher implements the TaskLauncher plugin interface to support
+// launching on-demand tasks for the Waypoint server.
+type TaskLauncher struct {
+	config TaskLauncherConfig
+}
+
+// StartTaskFunc implements component.TaskLauncher
+func (p *TaskLauncher) StartTaskFunc() interface{} {
+	return p.StartTask
+}
+
+// StopTaskFunc implements component.TaskLauncher
+func (p *TaskLauncher) StopTaskFunc() interface{} {
+	return p.StopTask
+}
+
+// TaskLauncherConfig is the configuration structure for the task plugin.
+type TaskLauncherConfig struct {
+	// Context specifies the kube context to use.
+	Context string `hcl:"context,optional"`
+
+	// KubeconfigPath is the path to the kubeconfig file. If this is
+	// blank then we default to the home directory. If we are running within
+	// a pod, we will use the service account authentication if available if
+	// this isn't set.
+	KubeconfigPath string `hcl:"kubeconfig,optional"`
+
+	// The name of the Kubernetes secret to use to pull images started by
+	// this task.
+	ImageSecret string `hcl:"image_secret,optional"`
+
+	// ServiceAccount is the name of the Kubernetes service account to apply to the
+	// application deployment. This is useful to apply Kubernetes RBAC to the pod.
+	ServiceAccount string `hcl:"service_account,optional"`
+
+	// Set an explicit pull policy for this task launching. By default
+	// we use "PullIfNotPresent" unless the image tag is "latest" when we
+	// use "Always".
+	PullPolicy string `hcl:"image_pull_policy,optional"`
+}
+
+func (p *TaskLauncher) Documentation() (*docs.Documentation, error) {
+	doc, err := docs.New(
+		docs.FromConfig(&TaskLauncherConfig{}),
+		docs.FromFunc(p.StartTaskFunc()),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	doc.Description(`
+Launch a Kubernetes pod for on-demand tasks from the Waypoint server.
+
+This will use the standard Kubernetes environment variables to source
+authentication information for Kubernetes. If this is running within Kubernetes
+itself (typical for a Kubernetes-based installation), it will use the pod's
+service account unless other auth is explicitly given. This allows the task
+launcher to work by default.
+`)
+
+	doc.Example(`
+task {
+	use "kubernetes" {}
+}
+`)
+
+	doc.SetField(
+		"kubeconfig",
+		"path to the kubeconfig file to use",
+		docs.Summary("by default uses from current user's home directory"),
+		docs.EnvVar("KUBECONFIG"),
+	)
+
+	doc.SetField(
+		"context",
+		"the kubectl context to use, as defined in the kubeconfig file",
+	)
+
+	doc.SetField(
+		"image_secret",
+		"name of the Kubernetes secrete to use for the image",
+		docs.Summary(
+			"this references an existing secret, waypoint does not create this secret",
```

Comment:
```suggestion
			"this references an existing secret; waypoint does not create this secret",
```

Nit: _I'm not totally sure_ my suggestion here is correct